### DISCUSSION TITLE
Problem in using user-defined covariance base class (CovModel)

### DISCUSSION THREAD

**Discussion by himansh78** [GeoStat-Framework/GSTools#194] [Q&A] [answered]
I am trying to generate a user-defined spatial random field in 2D (Earthquake slip realisations in my case) using the CovModel base class for Von-Karmen  Auto-correlation function ([link](https://reproducibility.org/RSF/book/sep/fractal/paper_html/node4.html) ). Unfortunately I am stuck at an error message and it is getting difficult to get past it.  It would be a great help if you could help me get some insights to solve the issue.

 

Here is the Covmodel I created-

```python
class VK(gs.CovModel):

    def cor(self, h):
        h = np.array(h, dtype=np.double)
        nu = 0.8
        res= (4*np.pi*nu*(h**nu)*sps.kv(nu, h)) /sps.kv(nu,h.min() )
        return res
```
I can plot spectral density using the covmodel but cannot create the model using SRF method.
```python
model = VK(dim=2)
slip_srf = gs.SRF(model,seed=10)
```

I get a TypeError when I run the above commands.- TypeError: 'float' object cannot be interpreted as an integer
I get the same error, even when I create a gaussian covariance model (as given in the tutorial itself).
Thanks a lot for your time. Looking forward to your response.

**Comment by MuellerSeb** [comment, marked as answer]
Hey there,
thanks for using GSTools.

I see two problems here:
1. the integer error comes from `emcee` but we fixed this in GSTools v1.3.2. so upgrading GSTools should solve this. (See: https://github.com/GeoStat-Framework/GSTools/issues/183)
2. the covariance model given with your link seems to be wrong, since they are dividing by `K(v,0)`, which is infinit and the model would result in a constantly zero one.
![wrong-cov.jpg](https://user-images.githubusercontent.com/19690642/126866869-f764f849-6edb-4a08-8144-f889bdbc0866.jpg)
The cited reference has a similar definition:
![Screenshot_20210724-131259.jpg](https://user-images.githubusercontent.com/19690642/126866888-b9c31754-4f7c-4f9a-8ca3-b4c5676014dd.jpg)
But they use another function with a finite value to normalize the function.
In the end, the described model is exatly the [matern model](https://geostat-framework.readthedocs.io/projects/gstools/en/stable/generated/gstools.covmodel.Matern.html#gstools.covmodel.Matern) with a slightly modified rescaling factor. So I would recommend using this predefined model.

Hope that helps!
Cheers, sebastian

**Reply by himansh78** [reply]
Thanks . that helps.
